test: cover base[32|58|64] with symmetric roundtrip fuzz (and padding) tests

l0rinc opened the pull request:
Added fuzzed roundtrips for `base[32|58|64]` encoding to make sure encoding/decoding are symmetric.
Note that if we omit the padding in `EncodeBase32` we won't be able to decode it with `DecodeBase32`.
Added dedicated padding tests to cover failure behavior
Also moved over the Base58 json test edge cases from https://github.com/bitcoin/bitcoin/pull/30035

l0rinc committed:
test: Add padding tests for Base32/Base64

fjahr commented on src/test/base32_tests.cpp:
```diff
@@ -31,9 +32,45 @@ BOOST_AUTO_TEST_CASE(base32_testvectors)
+    BOOST_CHECK( DecodeBase32("aaaaaaa="s));
+}
+
+BOOST_AUTO_TEST_CASE(base32_roundtrip) {
```
Shouldn't this be a fuzzing test?

l0rinc replied:
I've modelled these based on https://github.com/bitcoin/bitcoin/blob/master/src/test/base58_tests.cpp#L84-L98 - but I guess the existing fuzz tests already cover this as well, so I'll migrate them all to https://github.com/bitcoin/bitcoin/blob/master/src/test/fuzz/base_encode_decode.cpp#L20-L54, thanks for the hint

l0rinc replied:
Done

fjahr commented on src/test/base64_tests.cpp:
```diff
@@ -37,10 +38,42 @@ BOOST_AUTO_TEST_CASE(base64_testvectors)
+    BOOST_CHECK( DecodeBase64("YWE="s));
+}
+
+BOOST_AUTO_TEST_CASE(base64_roundtrip) {
```
Here too, this is probably better implemented as a fuzzing test.

l0rinc committed:
test: Fuzz Base32/Base58/Base64 roundtrip conversions

This commit introduces symmetric encode-decode roundtrips for all bases.
Minor refactors were also included:
• Split each base into a separate fuzz target.
• Added symmetric encode-decode roundtrip tests for all bases.
• Removed trim testing for encoded_string, as Base58 does not use whitespace padding.
• Made comparisons stricter by removing unnecessary lowercase conversions for bases that have mixed-case alphabets.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

l0rinc committed:
test: Extend base58_encode_decode.json with edge cases

Added edge cases, such as transitions at powers of 58, to verify boundary behavior.

l0rinc commented:
@maflcko, the Fuzz tests seem to be timing out now, is the new setup too heavy, should I maybe add them to different suites?

l0rinc commented:
> it seems odd that the fuzz tests would time out here

Split the big fuzz target into separate base58_encode_decode/base58check_encode_decode/base32_encode_decode/base64_encode_decode/psbt_base64_decode - it's cleaner like this anyway.

dergoegge commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -10,45 +10,86 @@
+        assert(encoded_string == ToLower(TrimString(random_string)));
     }
+    // Encode/Decode roundtrip
+    auto encoded{EncodeBase32(random_data)};
```
Reusing the same data from the fuzz input for two different things (e.g. decode/encode and encode/decode) is usually not a good idea.

How about adding a decode step after the encoding above (i.e. decode -> encode -> decode)? Or alternatively split this second round-trip into its own harness as well?

l0rinc replied:
Thanks for the review!

> is usually not a good idea

Can you please expand on that? I don't mind splitting, as you've recommended, but I have to understand what the current drawbacks are.

dergoegge replied:
It might be less obvious (and maybe not even a problem) for this simple harness but the decode/encode and encode/decode round-trips are different tests. Their input spaces are not the same, which means that reusing the same fuzz input for both leads to less efficient fuzzing (e.g. the corpus will include inputs that are only relevant to one of the tests but both will be executed).

l0rinc replied:
Thanks @maflcko.
@dergoegge, would a `CallOneOf` based solution satisfy your needs?

dergoegge replied:
> There is also an overhead that we probably don't want a separate fuzz target for every trivial function, because each fuzz target may be put into a separate binary and is allocated a separate folder.

I don't think the binary sizes would be a problem after we merge something like https://github.com/bitcoin/bitcoin/pull/30882 and the extra folder for the corpus is also not a huge cost (they compress really nicely).

I'd prefer to follow best practices over having to weigh triviality against some external overhead of running these tests. "Put separate tests into separate harnesses" is easier to follow than "aggregate trivial functions in one harness" because it is entirely subjective. What is trivial? How many unrelated trivial functions are allowed in one harness? I think this creates confusion for devs that aren't experts on fuzzing.

> would a CallOneOf based solution satisfy your needs?

I don't think this solves the issue, e.g. mutations that change the calloneof branch result in completely changing the meaning of the fuzz input (i.e. switching between the different tests).

I don't mean to derail this PR, it's already improving things by splitting up `base_encode_decode`. Feel free to leave it as is.

l0rinc replied:
Thanks for the directions @dergoegge, @maflcko.

> I don't mean to derail this PR, it's already improving things by splitting up base_encode_decode. Feel free to leave it as is.

I'm all for separation of concerns, that's what I've done here as well.
If there's a simple and non-controvertial way to make it even better, let me know!

hodlinator replied:
(My instinct would be to go with the following to avoid rocking the boat when it comes to seed corpus naming etc, but I'm too new to this aspect of fuzzing to carry much weight in this aspect).
```C++
void base58_encode_decode(FuzzBufferType buffer);
void base58check_encode_decode(FuzzBufferType buffer);
void base32_encode_decode(FuzzBufferType buffer);
void base64_encode_decode(FuzzBufferType buffer);
void psbt_base64_decode(FuzzBufferType buffer);

FUZZ_TARGET(base_encode_decode)
{
    base58_encode_decode(buffer);
    base58check_encode_decode(buffer);
    base32_encode_decode(buffer);
    base64_encode_decode(buffer);
    psbt_base64_decode(buffer);
}
```

l0rinc replied:
Is this a blocker for anyone? If anyone has strong feelings about this, please let me know, we can either restructure here or provide it in another PR. Resolving for now.

achow101 commented:
cc @hodlinator

hodlinator commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -10,45 +10,86 @@
+    if (DecodeBase58(random_string, decoded, 100)) {
+        auto encoded_string{EncodeBase58(decoded)};
+        assert(encoded_string == TrimString(random_string));
+        assert(encoded_string.empty() || !DecodeBase58(encoded_string, decoded, decoded.size() - 1));
```
Could test slightly more aggressively in line with the removed test:
https://github.com/bitcoin/bitcoin/blob/a5b4e42a575b4680a9eedb6ab2275d2a723cc20d/src/test/base58_tests.cpp#L93-L94

```suggestion
        assert(encoded_string.empty() || !DecodeBase58(encoded_string, decoded, fuzzed_data_provider.ConsumeIntegralInRange(0, decoded.size() - 1)));
```

l0rinc replied:
Good idea, this will simplify the string generation as well since it requires a `FuzzedDataProvider` anyway

hodlinator reviewed (commented):
Concept ACK ad917fb105b68fb369455ec7c42292a7673db1ad

- Extra tests/coverage, does not modify files containing the functionality being tested.
- Does remove `base58_random_encode_decode` unit test, but everything seems to be carried over to the modified fuzz test (except for testing for a variable count of insufficient result buffer sizes).

l0rinc commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -10,45 +10,86 @@
+FUZZ_TARGET(base58_encode_decode)
 {
-    const std::string random_encoded_string(buffer.begin(), buffer.end());
+    std::string random_string(buffer.begin(), buffer.end());
+    std::vector<unsigned char> random_data{ToByteVector(random_string)};
```
> Would avoid stripping const for these slightly longer lived variables, just for peace of mind.

Done, thanks.

> Not super happy about the memcmp though

agree, `memcmp` is ugly, but `std::ranges::equal` should work here.

l0rinc commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -10,45 +10,86 @@
     PartiallySignedTransaction psbt;
     std::string error;
-    (void)DecodeBase64PSBT(psbt, random_encoded_string, error);
+    (void) DecodeBase64PSBT(psbt, random_string, error);
```
Perfection! ✨👌✨

l0rinc commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -10,45 +10,86 @@
-    if (DecodeBase58(random_encoded_string, decoded, 100)) {
-        const std::string encoded_string = EncodeBase58(decoded);
-        assert(encoded_string == TrimStringView(encoded_string));
-        assert(ToLower(encoded_string) == ToLower(TrimString(random_encoded_string)));
```
astute observation ;)

hodlinator commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -6,49 +6,90 @@
 {
-    const std::string random_encoded_string(buffer.begin(), buffer.end());
+    FuzzedDataProvider provider(buffer.data(), buffer.size());
+    const std::string random_string{provider.ConsumeRemainingBytesAsString()};
```
I'm afraid `FuzzedDataProvider::remaining_bytes_` becomes zero after this line in the current version, meaning `provider.ConsumeIntegralInRange<int>(0, decoded.size() - 1)` below will always return the minimum value if I'm reading the `ConsumeIntegralInRange`-implementation correctly.

```suggestion
    const std::string random_string{provider.ConsumeRandomLengthString(101)};
```
---

Could maybe (in another PR) add something like `bool FuzzedDataProvider::consume_remaining_called_{false}` which is set by `ConsumeRemainingBytes*()` and then assert against it in the various consume methods, to catch cases like these.

l0rinc replied:
[Done](https://github.com/bitcoin/bitcoin/compare/5bc0a8daab37a826a59ed83eb59aef4081f1fe19..6fd185c035c1cc4dd961cf14a2087e97fb069440), thanks

hodlinator replied:
@dergoegge any thoughts on my suggestion of altering `FuzzedDataProvider` to trigger asserts when calls like `ConsumeIntegralInRange<int>(0, decoded.size() - 1)` are made after calls to `ConsumeRemainingBytes*()`? Maybe such an improvement could be upstreamed even (possibly with a flag to turn it off for sloppy tests).

hodlinator replied:
It might add extra friction to agree on preprocessor logic to make the code still compile on other compilers. Also, doing it at compile-time means it's an all/nothing thing that can't be turned off for individual sloppy tests. But I agree it would be great to see if upstream would take a compile-time version first.

hodlinator replied:
Yeah, hopefully shouldn't be much of an obstacle, maybe I'm too pessimistic about that aspect.

l0rinc replied:
Is there anything else for me to do here?

hodlinator replied:
Nothing more to do as far as this PR goes. Still interested to hear from folks more focused on fuzzing.

dergoegge replied:
Re https://github.com/bitcoin/bitcoin/pull/30746#discussion_r1824164226: I think I agree with marco that a compile time check would be nicer and we should ideally upstream it so that our `FuzzedDataProvider` impl stays in sync.

hodlinator replied:
Great that you made time to look into this!

> An alternative could be to forbid the direct use of the depleting member functions (with a linter) and add a new helper which eats a `FuzzedDataProvider&&` and returns the bytes in it. This way, the existing use-after-move checks can be used.

So you mean disallowing direct calls to `ConsumeRemainingBytes*`-family of methods only, and provide a helper that does only that part instead after swallowing up `FuzzedDataProvider&&` so it should no longer be used. Sounds like it's worth a try to me!

hodlinator reviewed (approved):
ACK 6fd185c035c1cc4dd961cf14a2087e97fb069440

range-diff shows all notable changes since my [Concept ACK](https://github.com/bitcoin/bitcoin/pull/30746#pullrequestreview-2397249971) are in response to my suggestions.

Ran each new fuzz target with:
```
₿ FUZZ=<target name> build_fuzz/src/test/fuzz/fuzz -max_total_time=30 -jobs=8
```
Also ran:
```
₿ ctest -j<X> --test-dir build --output-on-failure -R base
```

marcofleon reviewed (commented):
Code review ACK 6fd185c035c1cc4dd961cf14a2087e97fb069440

Nice improvements to the fuzz test. The separation into different targets for each encoding looks good to me. I guess in principle, each roundtrip should be separated as well. But seems fine as is for these simple targets.

hodlinator reviewed (approved):
re-ACK 6fd185c035c1cc4dd961cf14a2087e97fb069440

(Hoping Drahtbot will react properly to this comment.. already ACKed the same PR version above).

l0rinc commented on src/test/base64_tests.cpp:
```diff
@@ -37,10 +37,22 @@ BOOST_AUTO_TEST_CASE(base64_testvectors)
+    BOOST_CHECK(!DecodeBase64("invalid\0"s)); // correct size, invalid due to \0
+    BOOST_CHECK( DecodeBase64("nQB/pZw="s)); // valid
+    BOOST_CHECK(!DecodeBase64("nQB/pZw=\0invalid"s)); // correct size, invalid due to \0
+    BOOST_CHECK(!DecodeBase64("nQB/pZw=invalid\0"s)); // invalid size
```
I forgot about this comment, thanks for checking, both tests on 42 and 43 were invalid because of the non-trailing `=`, not the `\0` - [updated them](https://github.com/bitcoin/bitcoin/compare/6fd185c035c1cc4dd961cf14a2087e97fb069440..37e5fcfd4fad2f2daed041ddfd60b0c28e5da6b6)

hodlinator replied:
They are invalid, but still worth testing regardless of why precisely the current implementation errors out?

```C++
BOOST_CHECK(!DecodeBase64("nQB/pZw=\0invalid"); // padding and NUL followed by invalid data
BOOST_CHECK(!DecodeBase64("nQB/pZw=invalid\0"); // padding followed by invalid data and NUL
```

l0rinc replied:
What's the advantage of the comment stating exactly what the code already states?
We're already testing `\0` and non-trailing `=`, what else do these ones add?

hodlinator replied:
> What's the advantage of the comment stating exactly what the code already states? 

Valid point. Better to describe what is wrong according to the Base64 standard (+ `\0` I guess).

> We're already testing `\0` and non-trailing `=`, what else do these ones add?

They cover slightly different cases, mixing of `\0` and `=`. They may not test a unique error-handling path with the current implementation, but if they can uncover issues with a potential future implementation I think it was a mistake to remove them (@maflcko was just stating the comments were invalid as I understand it).

l0rinc replied:
Done, thanks

hodlinator commented on src/test/base64_tests.cpp:
```diff
@@ -37,10 +37,21 @@ BOOST_AUTO_TEST_CASE(base64_testvectors)
-    BOOST_CHECK(DecodeBase64("nQB/pZw="s));
-    BOOST_CHECK(!DecodeBase64("nQB/pZw=\0invalid"s));
-    BOOST_CHECK(!DecodeBase64("nQB/pZw=invalid\0"s));
+    BOOST_CHECK(!DecodeBase64("invalid\0"s)); // correct size, invalid due to \0
+    BOOST_CHECK( DecodeBase64("nQB/pZw="s)); // valid
+    BOOST_CHECK(!DecodeBase64("nQB/pZw=invalid"s)); // invalid, padding only allowed at the end
```
Last line no longer has an embedded NUL character as specified in the comment above this code block (`// Decoding strings with embedded NUL characters should fail`). Might belong more in `base64_padding` or at least in its own newline-separated block here:
```suggestion
    BOOST_CHECK(!DecodeBase64("invalid\0"s)); // correct size, invalid due to \0
    BOOST_CHECK( DecodeBase64("nQB/pZw="s)); // valid

    BOOST_CHECK(!DecodeBase64("nQB/pZw=invalid"s)); // invalid, padding only allowed at the end
```

l0rinc replied:
moved the values without \0 above the comment in 32 and 64 tests, thanks!

l0rinc commented on src/test/base32_tests.cpp:
```diff
@@ -31,9 +31,25 @@ BOOST_AUTO_TEST_CASE(base32_testvectors)
+    BOOST_CHECK_EQUAL(EncodeBase32("fooba"), "mzxw6ytb");
+
+    // Valid size
+    BOOST_CHECK(!DecodeBase32("========"s));
```
Done, thanks, kept the `"sv` only for the values with `\0` in them

l0rinc commented:
Thanks @maflcko & @hodlinator for the comments, I've [rebased](https://github.com/bitcoin/bitcoin/compare/37e5fcfd4fad2f2daed041ddfd60b0c28e5da6b6..7ae68c6ac2d25a7373fc4f8fc5241dd9fec744ec) the PR in the first push without any changes and [addresses your concerns](https://github.com/bitcoin/bitcoin/compare/7ae68c6ac2d25a7373fc4f8fc5241dd9fec744ec..3b66b7e20ddba94d251c97818700d46030b14cc5) in the second one - let me know if there's anything left:
* Separated `\0` tests from length validation
* Suffixed strings with `\0` with `"sv`, kept the rest as normal strings
* Brought back the `=\0` case

hodlinator reviewed (approved):
re-ACK 3b66b7e20ddba94d251c97818700d46030b14cc5

- Confirmed that changes since last review match https://github.com/bitcoin/bitcoin/pull/30746#issuecomment-2519739672. Thank you for incorporating our feedback.
- Ran unit tests and each fuzz target for 5 seconds.

l0rinc committed:
fuzz: Add fuzzing for max_ret_len in DecodeBase58/DecodeBase58Check

Different values are used for max_ret_len throughout the codebase (e.g., 21, 34, 78).
Theoretically, negative and zero values are also permitted. Let's stress-test those as well.

Co-authored-by: brunoerg <brunoely.gc@gmail.com>

hodlinator commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -6,49 +6,92 @@
 {
-    const std::string random_encoded_string(buffer.begin(), buffer.end());
+    FuzzedDataProvider provider(buffer.data(), buffer.size());
+    const std::string random_string{provider.ConsumeRandomLengthString(1000)};
+    const auto max_ret_len{provider.ConsumeIntegralInRange<int>(1, 300)};
```
nit: As `random_string` is currently `0-1000` long, and `max_ret_len` is `1-300`, most of the time `DecodeBase58()` is just returning `false` since `max_ret_len` is too small. But maybe that is intentional to stress that aspect?

Might be better to have more similar lengths, and also to include `0` as `max_ret_len` since it might occur in places where that parameter is calculated.

```suggestion
    const std::string random_string{provider.ConsumeRandomLengthString(1000)};
    const auto max_ret_len{provider.ConsumeIntegralInRange<int>(0, 1000)};
```

l0rinc replied:
Good point, and even negative values theoretically - thanks, [pushed](https://github.com/bitcoin/bitcoin/compare/a32002f2d5cff72639c9782d70ae52ec162d9ef8..6983d82c84d4ca351a7cd9d1e0e20ab878da6475)

hodlinator reviewed (approved):
re-ACK a32002f2d5cff72639c9782d70ae52ec162d9ef8

- Only added commit that fuzzes the max length parameter of `DecodeBase58`&`DecodeBase58Check`.

Re-ran the 2 modified fuzz-targets 30 seconds each.

hodlinator reviewed (approved):
re-ACK 6983d82c84d4ca351a7cd9d1e0e20ab878da6475

Responding to my previous feedback + testing negative max return values.

Retested modified targets.

TheCharlatan commented on src/test/base58_tests.cpp:
```diff
@@ -82,20 +82,4 @@ BOOST_AUTO_TEST_CASE(base58_DecodeBase58)
     BOOST_CHECK(!DecodeBase58Check("3vQB7B6MrGQZaxCuFg4oh\0" "0IOl"s, result, 100));
```
In commit 3b66b7e20ddba94d251c97818700d46030b14cc5

`s/Spit/Split` in commit message.

l0rinc replied:
Hah, indeed, proofread all other commit messages:
```bash
git range-diff 836d733a6788c7f7e66da6f8007fb62f15161429..6983d82c84d4ca351a7cd9d1e0e20ab878da6475 ae40cf1a8e16462a8b9dfd076d440bc8ec796c2b..f919d919eb8425ef2bb25aa0ebf61c90ab9b07fa
```

hodlinator commented:
re-ACK f919d919eb8425ef2bb25aa0ebf61c90ab9b07fa

Only commit message improvements since last ACK.

---

PR summary nit:

> Added dedicated padding tests to hardcode the failure behavior.

Not sure "hardcode" is the best verb there, how about:

"Added dedicated padding tests to cover failure behavior."?

achow101 commented:
f919d919eb8425ef2bb25aa0ebf61c90ab9b07fa

marcofleon commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -6,49 +6,92 @@
+        && std::ranges::equal(roundtrip_decoded, buffer));
+}
+
+FUZZ_TARGET(base58check_encode_decode)
```
Looking into this. Was there any reason to change `max_ret_len` from 100 to 1000?

l0rinc replied:
> issues.oss-fuzz.com/issues/397734700](https://issues.oss-fuzz.com/issues/397734700:

`Access is denied to this issue`

> Looking into this. Was there any reason to change max_ret_len from 100 to 1000?

The scenarios when `maxlength` doesn't restrict `random_string` seems like the more useful case. Given that the actual value was capped at `1000` already, I have aligned it to that.
Also, @brunoerg extended 100 to 300 in https://github.com/bitcoin/bitcoin/pull/31577/files#diff-c4025879a0b70c954d3ef668bcecd8a3e043ede42510e1e0e8f7e56b69bf7663R27, this was inspired by that change.
Given that `max_ret_len` throughout the codebase is <100 (e.g., 21, 34, 78), we should be safe with setting both to 100 as well if this turns out to be a bottleneck.

@maflcko, @marcofleon, do you think that would work?

l0rinc replied:
Pushed https://github.com/bitcoin/bitcoin/pull/31917, thanks @maflcko & @marcofleon.

l0rinc commented on src/test/fuzz/base_encode_decode.cpp:
```diff
@@ -6,49 +6,90 @@
-        assert(ToLower(encoded_string) == ToLower(TrimString(random_encoded_string)));
+    // Decode/Encode roundtrip
+    std::vector<unsigned char> decoded;
+    if (DecodeBase58Check(random_string, decoded, 100)) {
```
> how could a fuzz engine work around the checksum? 

I was wondering the same, wasn't aware of the `qa-assets` project. I will contribute a few values after I'll understand how it works exactly, thanks for bringing it to my attention!

> However, given the removed unit test, I am not sure if the coverage is identical.

You mean https://github.com/bitcoin/bitcoin/pull/30746/files#diff-98e64689e3af8a62a1ad571f8a6fa93e4955eb73a16db94c4df5adb0e7d0282dL85-L99?
I meant to move it over to the fuzz suite, not remove it - https://github.com/bitcoin/bitcoin/pull/30746/files#diff-c4025879a0b70c954d3ef668bcecd8a3e043ede42510e1e0e8f7e56b69bf7663R46-R57
Are you suggesting the new functionality doesn't cover everything the old one did?

l0rinc replied:
> checksum in the test vector, but the fuzz test does not

Is that the case? In both cases we were doing roundtrips. But in the code you're quoting it's also trying randomly - which, as you pointed out, usually fails. But the roundtrip part is still there at the end.

I've moved these randomized unit tests to the fuzz part since it was explicitly requested: https://github.com/bitcoin/bitcoin/pull/30746#discussion_r1736402266

l0rinc replied:
> Someone added fuzz inputs that pass the checksum (acceptable).

I have pushed https://github.com/bitcoin/bitcoin/pull/31917 which should resolve this (by randomly generating inputs that can be decoded).

> Someone checked that the fuzz engine can break the checksum (ideal).

I will also add a dedicated fuzzing harnesses to the corpus later.

l0rinc replied:
> Someone checked that the fuzz engine can break the checksum (ideal).

It probably can, eventually, but ran it overnight with:
```C++
    const int max_ret_len{provider.ConsumeIntegralInRange<int>(-1, 100)};
    if (std::vector<unsigned char> decoded; DecodeBase58Check(random_string, decoded, max_ret_len)) {
        throw "cracked!" + random_string;
    }
```
It ran 134 million executions and it didn't fail - my new PR should fix that, thanks for the observation
